daemon/k8s_watcher.go

```diff
@@ -1452,6 +1487,44 @@ func (d *Daemon) deleteCiliumNetworkPolicyV2(cnp *cilium_v2.CiliumNetworkPolicy)
 	}
 }
 
+func updateRuleAnnotations(ciliumV2Store cache.Store, cnp *cilium_v2.CiliumNetworkPolicy) error {
+
+	updatedCNPFromStore, err := getUpdatedCNPFromStore(ciliumV2Store, cnp)
+	if err != nil {
+		return err
+	}
+
+	// Make a copy since the rule is a pointer, and any of its fields
+	// which are also pointers could be modified outside of this
+	// function.
+	updatedCNPFromStoreCopy := updatedCNPFromStore.DeepCopy()
+
+	// Only update annotations for node on which this agent is running.
+	nodeName := node.GetName()
+	cnpNodeStatus := cnp.GetPolicyStatus(nodeName)
+	updatedCNPFromStoreNodeStatus := updatedCNPFromStoreCopy.GetPolicyStatus(nodeName)
+
+	// Do not update annotations if policy rule is not enforced on the node.
+	if !updatedCNPFromStoreNodeStatus.Enforcing {
+		return fmt.Errorf("policy has not been enforced; not updating annotations")
```

Comment:
You replaced the previous controller with this one. If this controller returns an error here, this will deadlock for ever as the "Enforcing" boolean is never set again to "true" (unless the rule is deleted and re-added)

Reply:
I can update this to unconditionally plumb the annotations; it will be up to consumers of the API to not only check the Annotations, but also whether the CNP is "OK". Is this OK?

One question: is this really a deadlock? The current code for updating the CNP runs with a `RunInterval` of zero:

https://github.com/cilium/cilium/blob/a454be550557f17826976fd1f61928df1888dc86/daemon/k8s_watcher.go#L1342-L1418

See https://github.com/cilium/cilium/blob/a454be550557f17826976fd1f61928df1888dc86/pkg/controller/controller.go#L58-L61

per the documentation of `pkg/controller`, if `RunInterval` is zero, retries won't occur:

https://github.com/cilium/cilium/blob/a454be550557f17826976fd1f61928df1888dc86/pkg/controller/controller.go#L58-L61

Use of `RunInterval` in the controller package: 
https://github.com/cilium/cilium/blob/a454be550557f17826976fd1f61928df1888dc86/pkg/controller/controller.go#L179-L233

So, the current code only runs the controller once for updating the status of this CNP, because the RunInterval is zero.

Reply:
@ianvernon the controller will not run only once since `runFunc` will not be set to false here https://github.com/cilium/cilium/blob/a454be550557f17826976fd1f61928df1888dc86/pkg/controller/controller.go#L201-L213

Reply:
@aanm ah, you are right.